pkg/client/client_venconn_test.go

```diff
@@ -0,0 +1,381 @@
+package client_test
+
+import (
+	"context"
+	"crypto/x509"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/jetstack/preflight/api"
+	"github.com/jetstack/preflight/pkg/client"
+
+	"github.com/jetstack/venafi-connection-lib/api/v1alpha1"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/rest"
+	ctrlruntime "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+// These are using envtest (slow) rather than a fake clientset (fast) because
+// controller-runtime's fake clientset doesn't support server-side apply [1] and
+// also because we want to create serviceaccount tokens, which isn't supported
+// by the fake clientset either.
+//
+// The goal is to test the following behaviors:
+//
+//   - VenafiConnection's `accessToken` works as expected with a fake Venafi
+//     Cloud server.
+//   - VenafiConnection's `apiKey` and `tpp` can't be used by the user.
+//   - NewVenConnClient's `trustedCAs` works as expected.
+//
+// [1] https://github.com/kubernetes-sigs/controller-runtime/issues/2341
+func TestVenConnClient_PostDataReadingsWithOptions(t *testing.T) {
+	t.Parallel()
+
+	t.Run("valid accessToken", run(testcase{
+		given: undent(`
+			apiVersion: jetstack.io/v1alpha1
+			kind: VenafiConnection
+			metadata:
+			  name: venafi-components
+			  namespace: venafi
+			spec:
+			  vcp:
+			    url: FAKE_VENAFI_CLOUD_URL
+			    accessToken:
+			      - secret:
+			          name: accesstoken
+			          fields: [accesstoken]`),
+		expectReadyCondMsg: "ea744d098c2c1c6044e4c4e9d3bf7c2a68ef30553db00f1714886cedf73230f1",
```

Comment:
That's a confusing Ready condition message! What does it mean? Where does it come from?

Reply:
Hey. I've looked at the official [Venafi Connection API reference](https://docs.venafi.cloud/vaas/k8s-components/c-vc-apiref/?h=reference+venafi+conne#venaficonnectionstatusconditionsindex) page and wasn't able to find where this is documented. I don't think it is documented anywhere.

I found an example of status [here](https://github.com/jetstack/venafi-connection-lib/pull/146). It looks like this:

```yaml
kind: VenafiConnection
status:
  conditions:
  - type: VenafiEnhancedIssuerReady
    status: "True"
    reason: Generated a token
    message: 8db2c15ace6f4c7b59138909b6b69d6caca69e2d308e695206b5e15ddaf28e81
    tokenValidUntil: "2123-10-02T19:57:36Z"
```

What (I think) happens is that venafi-connection-lib stores some form of hash in the `message` field... The `reason` is actually a `message` (`reason` should be TokenGenerated or something like that), and the `message` contains a non-readable string that should rather be in some custom `tokenHash` field. Funnily enough, the official Venafi Connection API reference says that `message` is a human-readable explanation:

> Message is a human readable description of the details of the last transition, complementing reason.

I don't think this quirk is a big deal, but IMO it should be clearly stated in the API reference or somewhere else that `message` contains an opaque string used by venafi-connection-lib to work, and that `reason` contains the actual human-readable string. For example, in case of error, it shows:

```yaml
kind: VenafiConnection
status:
  conditions:
    - type: VenafiEnhancedIssuerReady
    status: "False"
    reason: |
      connection is not ready yet (Venafi self-test failed): error authenticating
      with Venafi: vcert error: server error: server unavailable: Get "https://api.venafi.cloud/v1/useraccounts":
      net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    message: ""
    tokenValidUntil: ""
```

Reply:
Right now, the message field contains a hash of the token.
We can change this, there is no logic that depends on this behavior (only the human experience).